index.js

```diff
@@ -23,81 +23,64 @@ const globalPath = originalPath
   .join(pathSeparator)
 ;
 
-
-function runVersionCommand(command, callback) {
-  process.env.PATH = globalPath;
-
-  exec(command, (execError, stdout, stderr) => {
-    const commandDescription = JSON.stringify(command);
-
-    if (!execError) {
-      return callback(null, {
-        version: stdout,
-      });
-    }
-
-    if (
-      (execError.code === 127)
-      ||
-      (runningOnWindows && execError.message.includes("is not recognized"))
-     ) {
-      return callback(null, {
-        notfound: true,
-      });
-    }
-
-    // something went very wrong during execution
-    let errorMessage = `Command failed: ${commandDescription}`
-
-    if (stderr) {
-      errorMessage += `\n\nstderr:\n${stderr.toString().trim()}\n`;
-    }
-
-    errorMessage += `\n\noriginal error message:\n${execError.message}\n`;
-
-    return callback(new Error(errorMessage));
-  });
-
-  process.env.PATH = originalPath;
-}
-
-// Return object containing only keys that a program exists for and
-// something valid was given.
-function normalizeWanted(wanted) {
-  if (!wanted) {
-    return {};
-  }
-
-  // Validate keys
-  wanted = filterObject(wanted, Boolean);
-
-  // Normalize to strings
-  wanted = mapValues(wanted, String);
-
-  // Filter existing programs
-  wanted = filterObject(wanted, (_, key) => tools[key]);
-
-  return wanted;
-}
-
 module.exports = function check(wanted, callback) {
   // Normalize arguments
   if (typeof wanted === "function") {
     callback = wanted;
     wanted = null;
   }
 
-  wanted = normalizeWanted(wanted);
+  const options = { callback };
+
+  options.wanted = normalizeWanted(wanted);
 
-  const commands = mapValues(
+  options.commands = mapValues(
     (
-      Object.keys(wanted).length
-      ? filterObject(tools, (_, key) => wanted[key])
+      Object.keys(options.wanted).length
+      ? filterObject(tools, (_, key) => options.wanted[key])
       : tools
     ),
     ({ getVersion }) => ( runVersionCommand.bind(null, getVersion) )
   );
 
+  if (runningOnWindows) {
+    runForWindows(options);
+  } else {
+    run(options);
+  }
+}
+
+function runForWindows(options) {
+  // See and understand https://github.com/parshap/check-node-version/issues/35
+  // before trying to optimize this function
+  //
+  // `chcp` is used instead of `where` on account of its more extensive availablity
+  // chcp: MS-DOS 6.22+, Windows 95+; where: Windows 7+
+  //
+  // Plus, in order to be absolutely certain, the error message of `where` would still need evaluation.
+
+  exec("chcp", (error, stdout) => {
+    if (error) {
+      throw error;
+    }
+
+    const codepage = stdout.match(/\d+/)[0];
+
+    if (codepage === "65001" || codepage === "437") {
+      // need not switch codepage
+      return run(options);
+    }
+
+    // reset codepage before exiting
+    const finalCallback = options.callback;
+    options.callback = (...args) => exec(`chcp ${codepage}`, () => finalCallback(...args));
```

Comment:
Should we handle errors here?
```suggestion
    options.callback = (...args) => exec(`chcp ${codepage}`, err => {
      if (err) {
        throw err;
      }
      finalCallback(...args);
    });
```

Reply:
Good catch! These errors, unlikely as they are, should definitely be handled. Will do 👍